azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/message.py

```diff
@@ -0,0 +1,43 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+import logging
+import uuid
+
+logger = logging.getLogger(__name__)
+
+
+class Message(object):
+    def __init__(self, data, message_id=None, content_encoding=None, content_type=None):
+        """
+        :param data: The  data that constitutes the payload
+        :param message_id: A user-settable identifier for the message used for request-reply patterns.Format: A case-sensitive string (up to 128 characters long) of ASCII 7-bit alphanumeric characters + {'-', ':', '.', '+', '%', '_', '#', '*', '?', '!', '(', ')', ',', '=', '@', ';', '$', '''}.
+        :param content_encoding: Content encoding of the message data. Can be 'utf-8', 'utf-16' or 'utf-32'.
+        :param content_type: Content type property used to routes with the message body. Can be 'application/json'.
+        """
+        self.data = data
+        self.custom_properties = {}
+        self.lock_token = None
+        self.message_id = message_id if message_id else uuid.uuid4()
+        self.sequence_number = None
+        self.to = None
+        self.expiry_time_utc = None
+        self.enqueued_time = None
+        self.correlation_id = None
+        self.user_id = None
+        self.ack = None
+        self.content_encoding = content_encoding
+        self.content_type = content_type
+
+    def add_custom_property(self, key, value):
+        """
+        Adds the key-value pair to the collection of user defined properties. The strings 'Authorization'
+        and 'iothub-to' are considered as reserved key names and will cause the
+        function to log error.
+        """
+        if key == "Authorization" or key == "iothub-to":  # Not sure of this. Was part of Node code.
+            raise TypeError("The keys 'iothub-to' or 'Authorization' are not allowed")
```

Comment:
it's not clear to me that this is needed - it's transport specific or use-case specific if anything. you should either remove it or run it by the service and confirm that this is the right place because it happens on all protocols in all messages.
if we end up removing it, i don't see the value-add anymore of the whole add_custom_property method over accessing the dictionary directly.

Reply:
there is a requirement in Node that says `SRS_NODE_IOTHUB_PROPERTIES_07_004: [** If `itemKey` contains any of the reserved key names then the `add` function will return `false`. **]`. 

---
In reply to: [238908248](https://github.com/Azure/azure-iot-sdk-python-preview/pull/27#discussion_r238908248) [](ancestors = 238908248)

Reply:
yep and I think this is premature right now. it's protocol-specific and MQTT doesn't force any of that. we'll add this if/when we need this

---
In reply to: [239227272](https://github.com/Azure/azure-iot-sdk-python-preview/pull/27#discussion_r239227272) [](ancestors = 239227272,238908248)